### ISSUE TITLE
[RFC #2] About the priority of observers

### ISSUE THREAD

**Issue by iamolivinius** [secure-software-engineering/rose#238]
My work with patterns and the observer engine over the last couple of days made me wonder how and why we're using a priority field in observers. 

I came to the conclusion that `priority` are useful in two different cases:

    a) Two or more observers have exactly the same pattern
    b) Two or more observers have patterns that are strict subsets of each other

Case a) is trivial. Either you don't want this condition at all or you have two observers with different names but the same pattern and don't want to remove one of them (deactivating one observer).

Case b) seems to be the reason why `priority` was introduced in the first place. Unfortunately I wasn't able to find much details about how the `priority` value is derived. For me it looks a bit random or trail and error. One could argue that the patterns already implicate their priority. The CSS selectors that we're using as patterns come with a specified weight model [[1]](https://www.w3.org/TR/css3-selectors/#specificity). In short a more specific selector has a higher priority than a more generic one. 

I suggest in future ROSE development we should at least evaluate the performance between CSS specificity and the hand crafted priority values. 

Check out [[2]](https://specificity.keegan.st/) for a more visual representation of CSS specificity.

**Comment by apoller** [comment]
You are right, b) is the reason why we introduced the priority field. However, I'm open to replace it in case another metrics can compete with it. However, we should first gather some experience with it and Felix can later reconsider your proposal.

**Comment by dharma-guardian** [comment]
I like the suggestion to use CSS specifity. We should run some performance tests.

There is also another minor case c) for priority. Right now, if an observer hits, the engine stops testing the remaining patterns and therefore priority can be used for minimal performance optimization.
